File: src/main/java/com/worldscape/voronoi/VoronoiControlPoint.java
package com.worldscape.voronoi;

public class VoronoiControlPoint {
    private final String id;
    private int x;
    private int z;
    private int color;
    private float size;
    private float weight;
    private String label;
    private String terrainType;
    private boolean selected;
    private boolean visible;
    public static final int DEFAULT_COLOR = -11886849;
    private static final int SELECTED_COLOR = -43691;
    public static final float DEFAULT_SIZE = 8.0f;
    public static final float DEFAULT_WEIGHT = 1.0f;

    public VoronoiControlPoint(String id, int x, int z, int color, float size, float weight, String label, String terrainType) {
        this.id = id;
        this.x = x;
        this.z = z;
        this.color = color;
        this.size = Math.max(1.0f, size);
        this.weight = Math.max(0.0f, weight);
        this.label = label != null ? label : "";
        this.terrainType = terrainType != null ? terrainType : "";
        this.selected = false;
        this.visible = true;
    }

    public VoronoiControlPoint(String id, int x, int z, int color) {
        this(id, x, z, color, 8.0f, 1.0f, null, "");
    }

    public VoronoiControlPoint(int x, int z, String terrainType) {
        this("auto_" + x + "_" + z, x, z, -11886849, 8.0f, 1.0f, null, terrainType);
    }

    public VoronoiControlPoint(VoronoiControlPoint other) {
        this.id = other.id;
        this.x = other.x;
        this.z = other.z;
        this.color = other.color;
        this.size = other.size;
        this.weight = other.weight;
        this.label = other.label;
        this.terrainType = other.terrainType;
        this.selected = other.selected;
        this.visible = other.visible;
    }

    public String getId() {
        return this.id;
    }

    public int getX() {
        return this.x;
    }

    public int getZ() {
        return this.z;
    }

    public int getColor() {
        return this.selected ? -43691 : this.color;
    }

    public int getOriginalColor() {
        return this.color;
    }

    public float getSize() {
        return this.size;
    }

    public float getWeight() {
        return this.weight;
    }

    public String getLabel() {
        return this.label;
    }

    public String getTerrainType() {
        return this.terrainType;
    }

    public boolean isSelected() {
        return this.selected;
    }

    public void setX(int x) {
        this.x = x;
    }

    public void setZ(int z) {
        this.z = z;
    }

    public void setColor(int color) {
        this.color = color;
    }

    public void setSize(float size) {
        this.size = Math.max(1.0f, size);
    }

    public void setWeight(float weight) {
        this.weight = Math.max(0.0f, weight);
    }

    public void setLabel(String label) {
        this.label = label != null ? label : "";
    }

    public void setTerrainType(String terrainType) {
        this.terrainType = terrainType != null ? terrainType : "";
    }

    public void setSelected(boolean selected) {
        this.selected = selected;
    }

    public boolean isVisible() {
        return this.visible;
    }

    public void setVisible(boolean visible) {
        this.visible = visible;
    }

    public long squaredDistanceTo(int px, int pz) {
        long dx = (long)this.x - (long)px;
        long dz = (long)this.z - (long)pz;
        return dx * dx + dz * dz;
    }

    public double distanceTo(int px, int pz) {
        return Math.sqrt(this.squaredDistanceTo(px, pz));
    }

    public boolean isWithinRadius(int px, int pz, long radiusSq) {
        return this.squaredDistanceTo(px, pz) <= radiusSq;
    }

    public void toggleSelection() {
        this.selected = !this.selected;
    }

    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof VoronoiControlPoint)) {
            return false;
        }
        VoronoiControlPoint other = (VoronoiControlPoint)obj;
        return this.id.equals(other.id);
    }

    public int hashCode() {
        return this.id.hashCode();
    }

    public String toString() {
        return String.format("VoronoiControlPoint[id=%s, x=%d, z=%d, type=%s, weight=%.2f]", this.id, this.x, this.z, this.terrainType, Float.valueOf(this.weight));
    }
}

File: src/main/kotlin/com/worldscape/voronoi/VoronoiControlPoint.kt
package com.worldscape.voronoi

class VoronoiControlPoint(
    val id: String,
    var x: Int,
    var z: Int,
    @get:JvmName("getRawColor")
    var color: Int,
    var size: Float,
    var weight: Float,
    var label: String,
    var terrainType: String,
    @get:JvmName("isSelected")
    var selected: Boolean = false,
    @get:JvmName("isVisible")
    var visible: Boolean = true
) {
    companion object {
        @JvmField val DEFAULT_COLOR = -11886849
        @JvmField val DEFAULT_SIZE = 8.0f
        @JvmField val DEFAULT_WEIGHT = 1.0f
        private val SELECTED_COLOR = -43691
    }

    constructor(id: String, x: Int, z: Int, color: Int) :
        this(id, x, z, color, DEFAULT_SIZE, DEFAULT_WEIGHT, "", "")

    constructor(x: Int, z: Int, terrainType: String) :
        this("auto_${x}_$z", x, z, DEFAULT_COLOR, DEFAULT_SIZE, DEFAULT_WEIGHT, "", terrainType)

    constructor(other: VoronoiControlPoint) :
        this(other.id, other.x, other.z, other.color, other.size, other.weight,
             other.label, other.terrainType, other.selected, other.visible)

    fun getColor(): Int = if (selected) SELECTED_COLOR else color

    fun getOriginalColor(): Int = color

    fun distanceTo(px: Int, pz: Int): Double =
        Math.sqrt(squaredDistanceTo(px.toLong(), pz.toLong()).toDouble())

    fun squaredDistanceTo(px: Long, pz: Long): Long {
        val dx = x.toLong() - px
        val dz = z.toLong() - pz
        return dx * dx + dz * dz
    }

    fun isWithinRadius(px: Int, pz: Int, radiusSq: Long): Boolean =
        squaredDistanceTo(px.toLong(), pz.toLong()) <= radiusSq

    fun toggleSelection() { selected = !selected }

    override fun equals(other: Any?): Boolean =
        this === other || (other is VoronoiControlPoint && id == other.id)

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String =
        "VoronoiControlPoint[id=$id, x=$x, z=$z, type=$terrainType, weight=$weight]"
}